File: android/app/src/main/java/com/calcplus/calculator/core/lock/TokenRecorder.kt
package com.calcplus.calculator.core.lock

import com.calcplus.calculator.core.disguise.DisguiseEvent

/**
 * The host-owned entry buffer, generalized from `KeySequenceRecorder` over
 * opaque token IDs so every lock face shares one implementation of the length
 * and overflow rules (decisions §1.2).
 *
 * | Event                        | Buffer   | Overflow flag       |
 * |------------------------------|----------|---------------------|
 * | `Token`, buffer < 32         | append   | unchanged           |
 * | `Token`, buffer == 32        | unchanged| **set**             |
 * | `RemoveLast`, buffer empty   | no-op    | unchanged           |
 * | `RemoveLast`, buffer non-empty| pop last | **unchanged (sticky)** |
 * | `Clear`                      | empty    | reset               |
 * | `Commit`                     | taken    | taken, then reset   |
 *
 * The overflow flag is sticky through [removeLast] on purpose: once a 33rd
 * token has been seen the entry is unrecoverable by backspacing, and the only
 * recovery is [clear] (PIN pad: long-press ⌫) or a commit that fails as too
 * long. Nothing here is ever logged.
 */
class TokenRecorder {
    companion object {
        const val MAX_TOKENS = 32
    }

    private val _buffer = mutableListOf<String>()
    val buffer: List<String> get() = _buffer.toList()

    var overflowed = false
        private set

    fun record(token: String) {
        if (_buffer.size >= MAX_TOKENS) {
            overflowed = true
        } else {
            _buffer.add(token)
        }
    }

    /** Pops the last token. The overflow flag deliberately survives. */
    fun removeLast() {
        if (_buffer.isNotEmpty()) _buffer.removeAt(_buffer.size - 1)
    }

    fun clear() {
        _buffer.clear()
        overflowed = false
    }

    data class Commit(val tokens: List<String>, val overflowed: Boolean)

    /** Snapshots the buffer and the flag, then resets both. */
    fun takeCommit(): Commit {
        val commit = Commit(_buffer.toList(), overflowed)
        clear()
        return commit
    }

    /**
     * Applies one face event. Returns a [Commit] for — and only for —
     * [DisguiseEvent.Commit]; every other event returns null.
     */
    fun handle(event: DisguiseEvent): Commit? = when (event) {
        is DisguiseEvent.Token -> { record(event.id); null }
        DisguiseEvent.RemoveLast -> { removeLast(); null }
        DisguiseEvent.Clear -> { clear(); null }
        DisguiseEvent.Commit -> takeCommit()
    }
}

File: ios/SafeBox/Lock/TokenRecorder.swift
import Foundation

/// Host-owned buffer of token IDs, identical for every face (decisions §1.2).
/// Generalized from iteration 1's `PasscodeRecorder` over `[CalcKey]`.
///
/// | Event | Buffer | Overflow flag |
/// |---|---|---|
/// | `token`, buffer < 32 | append | unchanged |
/// | `token`, buffer == 32 | unchanged | set |
/// | `removeLast`, empty | no-op | unchanged |
/// | `removeLast`, non-empty | pop | **unchanged (sticky)** |
/// | `clear` | empty | reset |
/// | `commit` | taken, then empty | reset |
///
/// The flag is sticky through `removeLast` on purpose: once a 33rd token has
/// been seen the entry is unrecoverable by backspacing, and the only ways out
/// are `clear` or a commit that fails as TOO_LONG.
///
/// In memory only. Never logged, never persisted.
struct TokenRecorder: Sendable, Equatable {
    static let maxTokens = 32

    private(set) var buffer: [String] = []
    private(set) var overflowed = false

    mutating func record(_ token: String) {
        if buffer.count >= Self.maxTokens {
            overflowed = true
        } else {
            buffer.append(token)
        }
    }

    mutating func removeLast() {
        guard !buffer.isEmpty else { return }
        buffer.removeLast()
        // overflowed deliberately untouched.
    }

    mutating func clear() {
        buffer.removeAll()
        overflowed = false
    }

    /// Returns the buffered tokens and the overflow flag, then resets both.
    mutating func takeCommit() -> (tokens: [String], overflowed: Bool) {
        defer { clear() }
        return (buffer, overflowed)
    }

    /// Applies one face event. Returns the commit payload for `.commit` and
    /// `nil` for every other event — so a caller can drive the whole stream
    /// through this single entry point.
    mutating func apply(_ event: DisguiseEvent) -> (tokens: [String], overflowed: Bool)? {
        switch event {
        case .token(let id):
            record(id)
            return nil
        case .removeLast:
            removeLast()
            return nil
        case .clear:
            clear()
            return nil
        case .commit:
            return takeCommit()
        }
    }
}
